ciphersuite/registry.go

```diff
@@ -0,0 +1,160 @@
+	c, err := cr.get(raw.Name())
+	if err != nil {
+		return nil, xerrors.Errorf("cipher suite: %v", err)
+	}
+
+	pk, err := c.PublicKey(raw)
+	if err != nil {
+		return nil, xerrors.Errorf("unpacking: %v", err)
+	}
+
+	return pk, nil
+}
+
+// UnpackSecretKey takes generic cipher data and tries to convert it
+// into a secret key of the associated implementation. The cipher suite
+// must be registered beforehand.
+func (cr *Registry) UnpackSecretKey(raw *RawSecretKey) (SecretKey, error) {
+	c, err := cr.get(raw.Name())
+	if err != nil {
+		return nil, xerrors.Errorf("cipher suite: %v", err)
+	}
+
+	sk, err := c.SecretKey(raw)
+	if err != nil {
+		return nil, xerrors.Errorf("unpacking: %v", err)
+	}
+
+	return sk, nil
+}
+
+// UnpackSignature takes generic cipher data and tries to convert it
+// into a signature of the associated implementation. The cipher suite
+// must be registered beforehand.
+func (cr *Registry) UnpackSignature(raw *RawSignature) (Signature, error) {
+	c, err := cr.get(raw.Name())
+	if err != nil {
+		return nil, xerrors.Errorf("cipher suite: %v", err)
+	}
+
+	sig, err := c.Signature(raw)
+	if err != nil {
+		return nil, xerrors.Errorf("unpacking: %v", err)
+	}
+
+	return sig, nil
+}
+
+// WithContext executes the fn by passing the cipher suite that is assigned
+// to the nameable parameter.
+func (cr *Registry) WithContext(n Nameable, fn func(CipherSuite) error) error {
+	suite, err := cr.get(n.Name())
+	if err != nil {
+		return xerrors.Errorf("looking up cipher suite: %v", err)
+	}
+
+	return fn(suite)
+}
+
+// GenerateKeyPair returns a random secret key and its associated public key. This
+// function will panic in case of error which means the cipher suite should
+// be known and the default randomness should not trigger an error. If it
+// happens, that means something is wrong with the configuration.
+func (cr *Registry) GenerateKeyPair(name Name, reader io.Reader) (PublicKey, SecretKey, error) {
+	c, err := cr.get(name)
+	if err != nil {
+		return nil, nil, xerrors.Errorf("searching for cipher suite: %v", err)
+	}
+
+	pk, sk, err := c.GenerateKeyPair(reader)
+	if err != nil {
+		return nil, nil, xerrors.Errorf("generating key pair: %v", err)
+	}
+
+	return pk, sk, nil
+}
+
+// Sign takes a secret key and a message and produces a signature. It will
+// return an error if the signature is not known.
+func (cr *Registry) Sign(sk SecretKey, msg []byte) (Signature, error) {
+	c, err := cr.get(sk.Name())
+	if err != nil {
+		return nil, xerrors.Errorf("cipher suite: %v", err)
+	}
+
+	sig, err := c.Sign(sk, msg)
+	if err != nil {
+		return nil, xerrors.Errorf("signing: %v", err)
+	}
+
+	return sig, nil
+}
+
+// Verify takes a public key, a signature and a message and performs a verification
+// that will return an error if the signature does not match the message. It
+// will also return an error if the cipher suite is unknown.
+func (cr *Registry) Verify(pk PublicKey, sig Signature, msg []byte) error {
```

Comment:
just a thought, shouldn't we differentiate the case where the signature is incorrect from the one where an error occurred in the program? It feels a bit strange that the verification outputs an error while it only says that a signature is invalid.
Maybe it's better to rename the function `IsValid(...)` and return a `(bool, error)`. IDK